### PR TITLE
Improve rxm send/recv performance with HMEM in use

### PR DESCRIPTION
**Pull request by j-xiong**
Call ofi_hmem_host_register() on the buffers to be used as bounce buffer. This can greatly reduce the overhead of HMEM copy.

### CODE REVIEW

File: src/hmem_ze.c [hunk 1]
```diff
@@ -309,6 +319,26 @@ ze_result_t ofi_zeDeviceGetProperties(ze_device_handle_t hDevice,
 	return (*libze_ops.zeDeviceGetProperties)(hDevice, pDeviceProperties);
 }
 
+ze_result_t ofi_zeDriverGetExtensionFunctionAddress(ze_driver_handle_t hDriver,
+						    const char *name,
+						    void **ppFunctionAddress)
+{
+	return (*libze_ops.zeDriverGetExtensionFunctionAddress)(hDriver, name,
```
Comment (shefty) [inline src/hmem_ze.c:326]: What?  No period to end this sentence that is a function name?  :)

File: src/hmem_ze.c [hunk 2]
```diff
@@ -574,6 +598,12 @@ static int ze_hmem_dl_init(void)
 		goto err_dlclose;
 	}
 
+	libze_ops.zeDriverGetExtensionFunctionAddress = dlsym(libze_handle, "zeDriverGetExtensionFunctionAddress");
+	if (!libze_ops.zeDriverGetExtensionFunctionAddress) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find zeDriverGetExtensionFunctionAddress\n");
```
Comment (shefty) [inline src/hmem_ze.c:603]: line lengths

File: src/hmem_ze.c [hunk 3]
```diff
@@ -952,6 +999,35 @@ int *ze_hmem_get_dev_fds(int *nfds)
...
+
+	if (host_reg_enabled) {
+		ze_ret = ofi_zexDriverImportExternalPointer(driver, ptr, size);
+		if (ze_ret) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+				"Failed to import host memory: ptr %p size %zd",
+				ptr, size);
+		}
+	}
+	return FI_SUCCESS;
```
Comment (shefty) [inline src/hmem_ze.c:1014]: Is it okay to always return success even if we fail or the feature is disabled?

File: prov/util/src/util_buf.c [hunk 4]
```diff
@@ -62,6 +63,8 @@ static int ofi_bufpool_region_alloc(struct ofi_bufpool_region *buf_region)
 				buf_region->flags = OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_NONSHARED;
 				pool->alloc_size = alloc_size;
 				pool->region_size = pool->alloc_size - pool->entry_size;
+				if (host_rgister)
```
Comment (shefty) [inline prov/util/src/util_buf.c]: spelling error

File: prov/util/src/util_buf.c [hunk 5]
```diff
@@ -83,6 +86,8 @@ static int ofi_bufpool_region_alloc(struct ofi_bufpool_region *buf_region)
 		if (!ret) {
 			buf_region->flags = OFI_BUFPOOL_NONSHARED;
 			pool->region_size = pool->alloc_size - pool->entry_size;
+			if (host_rgister)
```
Comment (shefty) [inline prov/util/src/util_buf.c]: spelling

File: prov/util/src/util_buf.c [hunk 6]
```diff
@@ -94,15 +99,22 @@ static int ofi_bufpool_region_alloc(struct ofi_bufpool_region *buf_region)
 		pool->attr.alignment = ofi_get_aligned_size(pool->attr.alignment, page_size);
 	}
 
-	return ofi_memalign((void **) &buf_region->alloc_region,
+	ret = ofi_memalign((void **) &buf_region->alloc_region,
 				roundup_power_of_two(pool->attr.alignment),
 				pool->alloc_size);
+	if (!ret && host_rgister)
```
Comment (shefty) [inline prov/util/src/util_buf.c]: spelling

File: prov/util/src/util_buf.c [hunk 7]
```diff
@@ -94,15 +99,22 @@ static int ofi_bufpool_region_alloc(struct ofi_bufpool_region *buf_region)
...
+	return ret;
 }
 
 static void ofi_bufpool_region_free(struct ofi_bufpool_region *buf_region)
 {
 	int ret;
 	struct ofi_bufpool *pool = buf_region->pool;
+	bool host_register = pool->attr.flags & OFI_BUFPOOL_HMEM_COPY;
+
+	if (host_register)
```
Comment (shefty) [inline prov/util/src/util_buf.c:117]: at least your spelling is consistent

File: prov/rxm/src/rxm_ep.c [hunk 8]
```diff
@@ -262,7 +262,7 @@ static int rxm_ep_create_pools(struct rxm_ep *rxm_ep)
 	attr.free_fn = rxm_buf_close;
 	attr.init_fn = rxm_init_rx_buf;
 	attr.context = rxm_ep;
-	attr.flags = OFI_BUFPOOL_NO_TRACK;
+	attr.flags = OFI_BUFPOOL_NO_TRACK | OFI_BUFPOOL_HMEM_COPY;
```
Comment (shefty) [inline prov/rxm/src/rxm_ep.c:265]: This should only be set if hmem is actively enabled on the ep.

File: prov/rxm/src/rxm_ep.c [hunk 9]
```diff
@@ -298,7 +298,7 @@ static int rxm_multi_recv_pool_init(struct rxm_ep *rxm_ep)
 		.alloc_fn	= NULL,
 		.init_fn	= NULL,
 		.context	= rxm_ep,
-		.flags		= OFI_BUFPOOL_NO_TRACK,
+		.flags		= OFI_BUFPOOL_NO_TRACK | OFI_BUFPOOL_HMEM_COPY,
```
Comment (shefty) [inline prov/rxm/src/rxm_ep.c:301]: same -- we're taking a registration hit even if hmem is not used by the app.

File: prov/util/src/util_buf.c [hunk 10]
```diff
@@ -63,7 +64,7 @@ static int ofi_bufpool_region_alloc(struct ofi_bufpool_region *buf_region)
 				buf_region->flags = OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_NONSHARED;
 				pool->alloc_size = alloc_size;
 				pool->region_size = pool->alloc_size - pool->entry_size;
-				if (host_rgister)
+				if (host_register)
```
Comment (shefty) [inline prov/util/src/util_buf.c:67]: fixed in wrong patch

**Review by shefty — CHANGES_REQUESTED** [review]

File: src/hmem_ze.c (hunk 1 above)
Comment (j-xiong) [reply src/hmem_ze.c:326]: haha, I don't like that, either.

File: src/hmem_ze.c (hunk 2 above)
Comment (j-xiong) [reply src/hmem_ze.c:603]: I missed this one. Will fix.

File: src/hmem_ze.c (hunk 3 above)
Comment (j-xiong) [reply src/hmem_ze.c:1014]: Yes, on failure or disabled it acts like no-op.

File: prov/util/src/util_buf.c (hunk 4 above)
Comment (j-xiong) [reply prov/util/src/util_buf.c]: Thanks, will fix.

File: prov/util/src/util_buf.c (hunk 7 above)
Comment (j-xiong) [reply prov/util/src/util_buf.c:117]: :) that's why the wrongs one was not caught at compile time

File: prov/util/src/util_buf.c (hunk 10 above)
Comment (j-xiong) [reply prov/util/src/util_buf.c:67]: Good catch!

**Comment by j-xiong** [comment]
@shefty All comments should have been addressed in the updated commits.

**Review by shefty — APPROVED** [review]

**Comment by j-xiong** [comment]
Intel CI errors are `truncation error` in `fi_rdm_tagged_peek` and `address already in use` in `fi_recv_cancel`. Both are with the `net` provider and are unrelated to this PR.

**Comment by shefty** [comment]
Btw, we're investigating those failures. fi_rdm_tagged_peek will fail with both net and sockets providers, unsure why.  The fi_recv_cancel error is some problem at test startup.

**Pull request merged**